### PR TITLE
Change NDEF{Record, Message} to interface

### LINKED ISSUE #232: NFCReadingEvent has an attribute with a type of dictionary?
http://w3c.github.io/web-nfc/#dom-nfcreadingevent defines 

`interface NFCReadingEvent : Event {
  readonly attribute NDEFMessage message;
};
`

but https://heycam.github.io/webidl/#idl-dictionaries says

> Dictionaries must not be used as the type of an attribute or constant.

### PR DESCRIPTION
**Pull request by Honry**
Because dictionary could not be used as the type of an attribute.

Fixed #232

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1392,14 +1414,14 @@ <h2>The <dfn>NDEFRecordType</dfn> enum</h2>
...
+    typedef (DOMString or ArrayBuffer or NDEFMessageInit) NDEFMessageSource;
 
     [Constructor(), SecureContext, Exposed=Window]
     interface NFCWriter {
-      Promise&lt;void&gt; push(NDEFMessageSource message, optional NFCPushOptions options);
+      Promise&lt;void&gt; push(NDEFMessageSource message, optional NFCPushOptions options={});
     };
 
-    [Constructor(optional NFCReaderOptions options), SecureContext, Exposed=Window]
+    [Constructor(optional NFCReaderOptions options={}), SecureContext, Exposed=Window]
```
Comment (kenchris) [inline index.html:1424]: if you add ={} I don't think you need the optional any longer?

File: index.html [hunk 2]
```diff
@@ -1142,67 +1142,89 @@
...
+      };
+
+      dictionary NDEFRecordInit {
         NDEFRecordType recordType;
         USVString mediaType;
         NDEFRecordData data;
       };
     </pre>
     <p>
       The <dfn>NDEFRecordData</dfn> is a union type representing data types allowed
```
Comment (kenchris) [inline index.html:1202]: Shouldn't you write/define that in Web IDL then?

File: index.html [hunk 3]
```diff
@@ -1142,67 +1142,89 @@
...
     </p>
     <pre class="idl">
       typedef (DOMString or unrestricted double or ArrayBuffer or Dictionary) NDEFRecordData;
 
-      dictionary NDEFRecord {
+      [Constructor(NDEFRecordInit recordInit), Exposed=Window]
+      interface NDEFRecord {
+        readonly attribute NDEFRecordType recordType;
+        readonly attribute USVString mediaType;
+        NDEFRecordData data();
```
Comment (kenchris) [inline index.html:1192]: Wasn't the idea that we would have multiple getters depending on data type? like `json()`?

File: index.html (hunk 1 above)
Comment (Honry) [reply index.html:1424]: This was a new ReSpec error:
`WebIDL validation error: Optional dictionary arguments must have a default value of {}`

Do you have any solution?

File: index.html (hunk 3 above)
Comment (kenchris) [reply index.html:1192]: a bit like (copied from other spec)

```
[NewObject] Promise<any> json();
[NewObject] Promise<USVString> text();
[NewObject] Promise<ArrayBuffer> arrayBuffer();
```

File: index.html (hunk 1 above)
Comment (kenchris) [reply index.html:1424]: OK, maybe that is new, let's see what @marcoscaceres says

File: index.html (hunk 2 above)
Comment (Honry) [reply index.html:1202]: Eh... Sorry, I didn't get your point.  `NDEFRecordData` is already defined in [here](https://github.com/w3c/web-nfc/pull/240/files#diff-eacf331f0ffc35d4b482f1d15a887d3bR1186).

File: index.html (hunk 3 above)
Comment (Honry) [reply index.html:1192]: I saw you've already commented your concern on the original issue thread, let's just move to there.

**Comment by Honry** [comment]
@kenchris, many thanks for your comments! @zolkis, @leonhsl, please help review it as well. Thanks!

(@leonhsl is very busy on spec implementation, so he asked me to address the issue. :D )

File: index.html (hunk 1 above)
Comment (marcoscaceres) [reply index.html:1424]: Yeah, you still need the optional :)

File: index.html (hunk 3 above)
Comment (marcoscaceres) [reply index.html:1192]: Does this json() need to be async? If no, please use `default toJSON()` instead.

Comment (kenchris) [reply index.html:1192]: I am not sure. So non async methods doing potential conversion are called toSomething()?

Comment (marcoscaceres) [reply index.html:1192]: Yep. And the default toJSON() has special free behavior defined WebIDL. Not sure if it’s suitable, so double check.

Comment (kenchris) [reply index.html:1192]: We should have this in the design principles doc

**Review by zolkis — APPROVED** [review]
With the suggested change (`default toJSON()`) LGTM.

File: index.html (hunk 3 above)
Comment (leonhsl) [reply index.html:1192]: I assume the toJSON() discussion is for the solution below, right? That'd be our future improvement so can we just accept `NDEFRecordData data()` in this PR for now?

> [NewObject] Promise<any> json();
> [NewObject] Promise<USVString> text();
> [NewObject] Promise<ArrayBuffer> arrayBuffer();

Comment (kenchris) [reply index.html:1192]: Yes

**Comment by leonhsl** [comment]
lgtm, Thanks!

**Review by leonhsl — APPROVED** [review]

**Comment by Honry** [comment]
@zolkis, @kenchris, can we land this now? :) I will address the suggested improvement for `NDEFRecordData data()` in a follow-up.

**Pull request merged**